{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the config file\n",
    "with open('../config/config.json', 'r') as f:\n",
    "    config = json.load(f)\n",
    "\n",
    "file_path = config[\"data_loc\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " ## Datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inference batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The inference batch has 1097 observations and 4 columns.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Abstract</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>34902587</td>\n",
       "      <td>Detection of porcine circovirus type 3 DNA in ...</td>\n",
       "      <td>Porcine circovirus type 3 (PCV3) is regularly ...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>35451025</td>\n",
       "      <td>Imputation of non-genotyped F1 dams to improve...</td>\n",
       "      <td>This study investigated using imputed genotype...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>34859764</td>\n",
       "      <td>Proposed multidimensional pain outcome methodo...</td>\n",
       "      <td>Castration of male piglets in the United State...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>35143972</td>\n",
       "      <td>Nanostructured lipid carriers loaded with an a...</td>\n",
       "      <td>Alopecia is a condition associated with differ...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>34872491</td>\n",
       "      <td>Genome-wide expression of the residual lung re...</td>\n",
       "      <td>BACKGROUND: Acute or chronic irreversible resp...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PMID                                              Title  \\\n",
       "0  34902587  Detection of porcine circovirus type 3 DNA in ...   \n",
       "1  35451025  Imputation of non-genotyped F1 dams to improve...   \n",
       "2  34859764  Proposed multidimensional pain outcome methodo...   \n",
       "3  35143972  Nanostructured lipid carriers loaded with an a...   \n",
       "4  34872491  Genome-wide expression of the residual lung re...   \n",
       "\n",
       "                                            Abstract  Label  \n",
       "0  Porcine circovirus type 3 (PCV3) is regularly ...      0  \n",
       "1  This study investigated using imputed genotype...      0  \n",
       "2  Castration of male piglets in the United State...      0  \n",
       "3  Alopecia is a condition associated with differ...      0  \n",
       "4  BACKGROUND: Acute or chronic irreversible resp...      0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define file path\n",
    "file_name = \"test_unlabeled.tsv\"\n",
    "final_path = os.path.join(file_path, file_name) \n",
    "\n",
    "# Load tsv file\n",
    "inference_batch = pd.read_csv(final_path, sep='\\t')\n",
    "print(f\"The inference batch has {inference_batch.shape[0]} observations and {inference_batch.shape[1]} columns.\")\n",
    "inference_batch.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Training Corpus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape of the original dataset: (11278, 4) \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Abstract</th>\n",
       "      <th>Category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17179536</td>\n",
       "      <td>Variance component analysis of quantitative tr...</td>\n",
       "      <td>In a previous study, QTL for carcass compositi...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17177700</td>\n",
       "      <td>Single nucleotide polymorphism identification,...</td>\n",
       "      <td>Pituitary adenylate cyclase-activating polypep...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17129674</td>\n",
       "      <td>Genetic resistance to Sarcocystis miescheriana...</td>\n",
       "      <td>Clinical and parasitological traits of Sarcocy...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>17121599</td>\n",
       "      <td>Results of a whole-genome quantitative trait l...</td>\n",
       "      <td>A whole-genome quantitative trait locus (QTL) ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>17057239</td>\n",
       "      <td>Unexpected high polymorphism at the FABP4 gene...</td>\n",
       "      <td>Fatty acid bing protein 4 (FABP4) plays a key ...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PMID                                              Title  \\\n",
       "0  17179536  Variance component analysis of quantitative tr...   \n",
       "1  17177700  Single nucleotide polymorphism identification,...   \n",
       "2  17129674  Genetic resistance to Sarcocystis miescheriana...   \n",
       "3  17121599  Results of a whole-genome quantitative trait l...   \n",
       "4  17057239  Unexpected high polymorphism at the FABP4 gene...   \n",
       "\n",
       "                                            Abstract  Category  \n",
       "0  In a previous study, QTL for carcass compositi...         1  \n",
       "1  Pituitary adenylate cyclase-activating polypep...         0  \n",
       "2  Clinical and parasitological traits of Sarcocy...         0  \n",
       "3  A whole-genome quantitative trait locus (QTL) ...         1  \n",
       "4  Fatty acid bing protein 4 (FABP4) plays a key ...         0  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define file path\n",
    "file_name = \"QTL_text.json\"\n",
    "final_path = os.path.join(file_path, file_name) \n",
    "\n",
    "# Load json file\n",
    "df = pd.read_json(final_path)\n",
    "df = df.drop(columns=['Journal'])\n",
    "print(f\"Shape of the original dataset: {df.shape}\", \"\\n\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pre-Processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package punkt to /Users/gabrielvictorgomesferr\n",
      "[nltk_data]     eira/opt/anaconda3/envs/nlp_env/nltk_data...\n",
      "[nltk_data]   Package punkt is already up-to-date!\n",
      "[nltk_data] Downloading package stopwords to /Users/gabrielvictorgomes\n",
      "[nltk_data]     ferreira/opt/anaconda3/envs/nlp_env/nltk_data...\n",
      "[nltk_data]   Package stopwords is already up-to-date!\n"
     ]
    }
   ],
   "source": [
    "import spacy\n",
    "nlp_spacy = spacy.load(\"en_core_web_sm\")\n",
    "\n",
    "import nltk\n",
    "nltk.download('punkt')\n",
    "nltk.download('stopwords')\n",
    "from nltk.tokenize import word_tokenize\n",
    "from nltk.corpus import stopwords, wordnet\n",
    "from nltk.stem import WordNetLemmatizer\n",
    "import string\n",
    "import re\n",
    "\n",
    "import gensim\n",
    "from gensim.models.phrases import Phraser, Phrases\n",
    "\n",
    "from sklearn.model_selection import train_test_split, cross_val_score\n",
    "from sklearn.metrics import mean_absolute_error, r2_score, make_scorer, f1_score, recall_score, precision_score\n",
    "\n",
    "\n",
    "# Set stop words\n",
    "stop_words = set(stopwords.words('english'))\n",
    "\n",
    "# Set lemmatizer\n",
    "lemmatizer = WordNetLemmatizer()\n",
    "\n",
    "# Defining NLP Pre-Processing steps. These steps were the one that highlighted the most on my exploration phase\n",
    "def nlp_preprocessing(abstract_tokenized):\n",
    "    abstract_tokenized = abstract_tokenized.apply(lambda doc: doc.lower()) # Lowercasing txt\n",
    "    abstract_tokenized = abstract_tokenized.apply(lambda doc: doc.translate(str.maketrans(\"\", \"\", string.punctuation))) # Removing punctuations from the text\n",
    "    # abstract_tokenized = abstract_tokenized.apply(lambda doc: re.sub(r\"\\d+\", \"\", doc)) # Removing numbers from the text\n",
    "    abstract_tokenized = abstract_tokenized.apply(lambda doc: [token for token in word_tokenize(doc) if token not in stop_words]) # Tokenizing and removing stop words from the text\n",
    "    abstract_tokenized = abstract_tokenized.apply(lambda doc: [lemmatizer.lemmatize(token, pos=wordnet.NOUN) for token in doc]) # Convert words to their base\n",
    "    abstract_tokenized = abstract_tokenized.apply(lambda doc: [token.strip() for token in doc if token.strip() and len(token)>1]) # Removing extra space\n",
    "    return abstract_tokenized\n",
    "\n",
    "# Function to join token lists into a string\n",
    "def tokens_to_text(tokens_series):\n",
    "    return tokens_series.apply(lambda tokens: \" \".join(tokens))\n",
    "\n",
    "# Key-phrase extraction\n",
    "def gensim_key_phrase_extractor(list_of_tokens):\n",
    "    bigram = Phraser(Phrases(list_of_tokens, min_count=2, threshold=15))\n",
    "    bigram_token = [bigram[doc] for doc in list_of_tokens]\n",
    "    return bigram_token\n",
    "\n",
    "# Define a helper function to run experiments via cross-validation.\n",
    "def run_experiment(X_text, y, pipeline, cv=5):\n",
    "    scores = cross_val_score(pipeline, X_text, y, cv=cv, scoring='f1_macro')\n",
    "    return scores.mean()\n",
    "\n",
    "# Evaluate Classification Model\n",
    "def evaluate_classification_model(y_true, y_pred):\n",
    "    results = []\n",
    "\n",
    "    # Evaluate model\n",
    "    md_recall_score = round(recall_score(y_true, y_pred, average='macro'), 5)\n",
    "    md_precision_score = round(precision_score(y_true, y_pred, average='macro'), 5)\n",
    "    md_f1_score = round(f1_score(y_true, y_pred, average='macro'), 5)\n",
    "\n",
    "    results.append({\n",
    "            \"recall_score\": md_recall_score,\n",
    "            \"precision_score\": md_precision_score,\n",
    "            \"f1_score\": md_f1_score\n",
    "            })\n",
    "\n",
    "    # Transform results into DF\n",
    "    results_df = pd.DataFrame(results)\n",
    "\n",
    "    return results_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Abstract</th>\n",
       "      <th>Category</th>\n",
       "      <th>title_nltk_token</th>\n",
       "      <th>abstract_nltk_token</th>\n",
       "      <th>title_bigram</th>\n",
       "      <th>abstract_bigram</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17179536</td>\n",
       "      <td>Variance component analysis of quantitative tr...</td>\n",
       "      <td>In a previous study, QTL for carcass compositi...</td>\n",
       "      <td>1</td>\n",
       "      <td>[variance, component, analysis, quantitative, ...</td>\n",
       "      <td>[previous, study, qtl, carcass, composition, m...</td>\n",
       "      <td>[variance_component, analysis, quantitative_tr...</td>\n",
       "      <td>[previous_study, qtl, carcass_composition, mea...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17177700</td>\n",
       "      <td>Single nucleotide polymorphism identification,...</td>\n",
       "      <td>Pituitary adenylate cyclase-activating polypep...</td>\n",
       "      <td>0</td>\n",
       "      <td>[single, nucleotide, polymorphism, identificat...</td>\n",
       "      <td>[pituitary, adenylate, cyclaseactivating, poly...</td>\n",
       "      <td>[single_nucleotide, polymorphism, identificati...</td>\n",
       "      <td>[pituitary, adenylate, cyclaseactivating, poly...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17129674</td>\n",
       "      <td>Genetic resistance to Sarcocystis miescheriana...</td>\n",
       "      <td>Clinical and parasitological traits of Sarcocy...</td>\n",
       "      <td>0</td>\n",
       "      <td>[genetic, resistance, sarcocystis, miescherian...</td>\n",
       "      <td>[clinical, parasitological, trait, sarcocystis...</td>\n",
       "      <td>[genetic, resistance, sarcocystis, miescherian...</td>\n",
       "      <td>[clinical, parasitological, trait, sarcocystis...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>17121599</td>\n",
       "      <td>Results of a whole-genome quantitative trait l...</td>\n",
       "      <td>A whole-genome quantitative trait locus (QTL) ...</td>\n",
       "      <td>1</td>\n",
       "      <td>[result, wholegenome, quantitative, trait, loc...</td>\n",
       "      <td>[wholegenome, quantitative, trait, locus, qtl,...</td>\n",
       "      <td>[result, wholegenome, quantitative_trait, locu...</td>\n",
       "      <td>[wholegenome, quantitative_trait, locus_qtl, s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>17057239</td>\n",
       "      <td>Unexpected high polymorphism at the FABP4 gene...</td>\n",
       "      <td>Fatty acid bing protein 4 (FABP4) plays a key ...</td>\n",
       "      <td>0</td>\n",
       "      <td>[unexpected, high, polymorphism, fabp4, gene, ...</td>\n",
       "      <td>[fatty, acid, bing, protein, fabp4, play, key,...</td>\n",
       "      <td>[unexpected, high, polymorphism, fabp4_gene, u...</td>\n",
       "      <td>[fatty_acid, bing, protein_fabp4, play_key, ro...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PMID                                              Title  \\\n",
       "0  17179536  Variance component analysis of quantitative tr...   \n",
       "1  17177700  Single nucleotide polymorphism identification,...   \n",
       "2  17129674  Genetic resistance to Sarcocystis miescheriana...   \n",
       "3  17121599  Results of a whole-genome quantitative trait l...   \n",
       "4  17057239  Unexpected high polymorphism at the FABP4 gene...   \n",
       "\n",
       "                                            Abstract  Category  \\\n",
       "0  In a previous study, QTL for carcass compositi...         1   \n",
       "1  Pituitary adenylate cyclase-activating polypep...         0   \n",
       "2  Clinical and parasitological traits of Sarcocy...         0   \n",
       "3  A whole-genome quantitative trait locus (QTL) ...         1   \n",
       "4  Fatty acid bing protein 4 (FABP4) plays a key ...         0   \n",
       "\n",
       "                                    title_nltk_token  \\\n",
       "0  [variance, component, analysis, quantitative, ...   \n",
       "1  [single, nucleotide, polymorphism, identificat...   \n",
       "2  [genetic, resistance, sarcocystis, miescherian...   \n",
       "3  [result, wholegenome, quantitative, trait, loc...   \n",
       "4  [unexpected, high, polymorphism, fabp4, gene, ...   \n",
       "\n",
       "                                 abstract_nltk_token  \\\n",
       "0  [previous, study, qtl, carcass, composition, m...   \n",
       "1  [pituitary, adenylate, cyclaseactivating, poly...   \n",
       "2  [clinical, parasitological, trait, sarcocystis...   \n",
       "3  [wholegenome, quantitative, trait, locus, qtl,...   \n",
       "4  [fatty, acid, bing, protein, fabp4, play, key,...   \n",
       "\n",
       "                                        title_bigram  \\\n",
       "0  [variance_component, analysis, quantitative_tr...   \n",
       "1  [single_nucleotide, polymorphism, identificati...   \n",
       "2  [genetic, resistance, sarcocystis, miescherian...   \n",
       "3  [result, wholegenome, quantitative_trait, locu...   \n",
       "4  [unexpected, high, polymorphism, fabp4_gene, u...   \n",
       "\n",
       "                                     abstract_bigram  \n",
       "0  [previous_study, qtl, carcass_composition, mea...  \n",
       "1  [pituitary, adenylate, cyclaseactivating, poly...  \n",
       "2  [clinical, parasitological, trait, sarcocystis...  \n",
       "3  [wholegenome, quantitative_trait, locus_qtl, s...  \n",
       "4  [fatty_acid, bing, protein_fabp4, play_key, ro...  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Applying pre-defined pre-processing steps\n",
    "df_processed = df.copy()\n",
    "df_processed['title_nltk_token'] = nlp_preprocessing(df_processed['Title'])\n",
    "df_processed['abstract_nltk_token'] = nlp_preprocessing(df_processed['Abstract'])\n",
    "\n",
    "# Extracting Bigrams\n",
    "df_processed['title_bigram'] = gensim_key_phrase_extractor(df_processed['title_nltk_token'])\n",
    "df_processed['abstract_bigram'] = gensim_key_phrase_extractor(df_processed['abstract_nltk_token'])\n",
    "\n",
    "df_processed.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Train-Test Split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define predictor and target features\n",
    "X = df_processed.drop(columns=['Category'])\n",
    "y = df_processed['Category']\n",
    "\n",
    "# Split train and test\n",
    "X_train, X_val, y_train, y_val = train_test_split(X,y, test_size=.2, random_state=42, stratify=y)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.svm import LinearSVC\n",
    "\n",
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import accuracy_score, classification_report, confusion_matrix\n",
    "from sklearn.preprocessing import MinMaxScaler, MaxAbsScaler\n",
    "\n",
    "# Normalize Overal Score\n",
    "scaler = MinMaxScaler()\n",
    "\n",
    "# Set different corpus representations\n",
    "def set_experiment_features(df):\n",
    "    experiments = {\n",
    "        \"title\": tokens_to_text(df['title_bigram']),\n",
    "        \"title raw\": df['Title'],\n",
    "        \"abstract\": tokens_to_text(df['abstract_bigram']),\n",
    "        \"abstract raw\": df['Abstract'],\n",
    "        \"title abstract\": tokens_to_text(df['title_bigram']) + \" \" + tokens_to_text(X_train['abstract_bigram']),\n",
    "        \"title abstract raw\": df['Title'] + \" \" + df['Abstract']\n",
    "    }\n",
    "    return experiments\n",
    "\n",
    "# Set different models to experiment\n",
    "def set_experiment_models():\n",
    "\n",
    "    model_pipelines = {\n",
    "        \"Logistic Regression\": Pipeline([\n",
    "            ('tfidf', TfidfVectorizer(stop_words='english')),\n",
    "            ('scale', MaxAbsScaler()),\n",
    "            ('clf', LogisticRegression(solver='liblinear', random_state=42))\n",
    "        ]),\n",
    "        \"Linear SVM\": Pipeline([\n",
    "            ('tfidf', TfidfVectorizer(stop_words='english')),\n",
    "            ('scale', MaxAbsScaler()),\n",
    "            ('clf', LinearSVC(random_state=42))\n",
    "        ])\n",
    "    }\n",
    "\n",
    "    return model_pipelines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Corpus: title, Model: Logistic Regression, CV F1 Macro: 0.803\n",
      "Corpus: title, Model: Linear SVM, CV F1 Macro: 0.817\n",
      "Corpus: title raw, Model: Logistic Regression, CV F1 Macro: 0.826\n",
      "Corpus: title raw, Model: Linear SVM, CV F1 Macro: 0.829\n",
      "Corpus: abstract, Model: Logistic Regression, CV F1 Macro: 0.850\n",
      "Corpus: abstract, Model: Linear SVM, CV F1 Macro: 0.859\n",
      "Corpus: abstract raw, Model: Logistic Regression, CV F1 Macro: 0.853\n",
      "Corpus: abstract raw, Model: Linear SVM, CV F1 Macro: 0.855\n",
      "Corpus: title abstract, Model: Logistic Regression, CV F1 Macro: 0.854\n",
      "Corpus: title abstract, Model: Linear SVM, CV F1 Macro: 0.861\n",
      "Corpus: title abstract raw, Model: Logistic Regression, CV F1 Macro: 0.860\n",
      "Corpus: title abstract raw, Model: Linear SVM, CV F1 Macro: 0.862\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Corpus</th>\n",
       "      <th>Model</th>\n",
       "      <th>CV_F1_macro</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>title abstract raw</td>\n",
       "      <td>Linear SVM</td>\n",
       "      <td>0.861863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>title abstract</td>\n",
       "      <td>Linear SVM</td>\n",
       "      <td>0.860781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>title abstract raw</td>\n",
       "      <td>Logistic Regression</td>\n",
       "      <td>0.860059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>abstract</td>\n",
       "      <td>Linear SVM</td>\n",
       "      <td>0.858711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>abstract raw</td>\n",
       "      <td>Linear SVM</td>\n",
       "      <td>0.854757</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Corpus                Model  CV_F1_macro\n",
       "0  title abstract raw           Linear SVM     0.861863\n",
       "1      title abstract           Linear SVM     0.860781\n",
       "2  title abstract raw  Logistic Regression     0.860059\n",
       "3            abstract           Linear SVM     0.858711\n",
       "4        abstract raw           Linear SVM     0.854757"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiments_train = set_experiment_features(X_train)\n",
    "model_pipelines = set_experiment_models()\n",
    "\n",
    "# Prepare a results list to collect performance metrics.\n",
    "results = []\n",
    "\n",
    "# Loop through each corpus and each model\n",
    "for corpus_name, corpus in experiments_train.items():\n",
    "    for model_name, pipeline in model_pipelines.items():\n",
    "         score = run_experiment(corpus, y_train, pipeline)\n",
    "         results.append({\n",
    "             \"Corpus\": corpus_name,\n",
    "             \"Model\": model_name,\n",
    "             \"CV_F1_macro\": score\n",
    "         })\n",
    "         print(f\"Corpus: {corpus_name}, Model: {model_name}, CV F1 Macro: {score:.3f}\")\n",
    "\n",
    "# Convert results to a DataFrame for a clear overview\n",
    "results_df = pd.DataFrame(results).sort_values('CV_F1_macro', ascending=False).reset_index(drop=True)\n",
    "results_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Validation Set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validation Performance:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Corpus</th>\n",
       "      <th>recall_score</th>\n",
       "      <th>precision_score</th>\n",
       "      <th>f1_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Linear SVM</td>\n",
       "      <td>title abstract raw</td>\n",
       "      <td>0.86486</td>\n",
       "      <td>0.8953</td>\n",
       "      <td>0.87931</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Model              Corpus  recall_score  precision_score  f1_score\n",
       "0  Linear SVM  title abstract raw       0.86486           0.8953   0.87931"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_method = results_df['Corpus'][0]\n",
    "\n",
    "# Create different corpus representations:\n",
    "experiments_val = set_experiment_features(X_val)\n",
    "\n",
    "X_train_corpurs = experiments_train[best_method]\n",
    "X_val_corpurs = experiments_val[best_method]\n",
    "\n",
    "best_model = results_df['Model'][0]\n",
    "best_pipeline = model_pipelines[best_model]\n",
    "\n",
    "# Train the pipeline on the training data\n",
    "best_pipeline.fit(X_train_corpurs, y_train)\n",
    "\n",
    "# Predict on the validation set\n",
    "y_pred = best_pipeline.predict(X_val_corpurs)\n",
    "\n",
    "# Evaluate the predictions\n",
    "evaluation_df = evaluate_classification_model(y_val, y_pred)\n",
    "evaluation_df.insert(0, 'Model', best_model)\n",
    "evaluation_df.insert(1, 'Corpus', best_method)\n",
    "\n",
    "print(\"Validation Performance:\")\n",
    "evaluation_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Corpus</th>\n",
       "      <th>recall_score</th>\n",
       "      <th>precision_score</th>\n",
       "      <th>f1_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Linear SVM</td>\n",
       "      <td>title abstract raw</td>\n",
       "      <td>0.86486</td>\n",
       "      <td>0.8953</td>\n",
       "      <td>0.87931</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Model              Corpus  recall_score  precision_score  f1_score\n",
       "0  Linear SVM  title abstract raw       0.86486           0.8953   0.87931"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_path = 'experiment_outputs'\n",
    "os.makedirs(output_path, exist_ok=True)\n",
    "\n",
    "exp_results_path = os.path.join(output_path, 'experiment_results.csv')\n",
    "                                \n",
    "experiments_history = pd.read_csv(exp_results_path)\n",
    "experiments_history.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Corpus</th>\n",
       "      <th>recall_score</th>\n",
       "      <th>precision_score</th>\n",
       "      <th>f1_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Linear SVM</td>\n",
       "      <td>title abstract raw</td>\n",
       "      <td>0.86486</td>\n",
       "      <td>0.8953</td>\n",
       "      <td>0.87931</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Model              Corpus  recall_score  precision_score  f1_score\n",
       "0  Linear SVM  title abstract raw       0.86486           0.8953   0.87931"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiments_history = pd.concat([experiments_history, evaluation_df], axis=0).sort_values('f1_score', ascending=False).reset_index(drop=True)\n",
    "experiments_history"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export Experiment Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "experiments_history.to_csv(exp_results_path, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nlp_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
